### DISCUSSION TITLE
[Question] Is UFCS supposed to use ADL?

### DISCUSSION THREAD

**Discussion by guybrush77** [hsutter/cppfront#681] [Q&A] [answered]
Sorry if this had been asked before, but is UFCS supposed to use ADL? For example:

```
foo: namespace = {
    print: (i: int) = std::cout << i << "\n";
}

main: () = {
    i := 42;
    i.print();
}
```

**Comment by JohelEGP** [comment, marked as answer]
Depends on what you mean by "suppose".
It can use ADL.
But it doesn't have to.
AFAIK, UFCS is simply defined as `𝘰𝘣𝘫.𝘧(𝘢𝘳𝘨𝘴)` if that is well-formed, and `𝘧(𝘰𝘣𝘫, 𝘢𝘳𝘨𝘴)` otherwise.
So by definition of the Cpp1 semantics of (non-member) function call, it can use ADL.

**Reply by guybrush77** [reply]
I see. In that case, the example above  will not work because, under current rules, the compiler won't try to generate `foo::print(i)`.

Perhaps this is a naive question, but if we had true ADL lookup in the core language, could this be made to work? Though perhaps it might open a can of worms...

**Reply by JohelEGP** [reply]
Yes, because that's not [ADL](https://en.cppreference.com/w/cpp/language/adl).
Fundamental types don't have an associated namespace.
